SleepNumberController_App.groovy

```diff
@@ -1807,68 +2324,266 @@ def httpRequest(String path, Closure method = this.&get, Map body = null, Map qu
+  Integer lim=(Integer)settings.limitErrorLogsMin
+  if (!lim /* off */
+      || (wnow() - lastErrorLogTime) > (lim * 60 * 1000)) {
+    logError msg,e
+    lastErrorLogTime = wnow()
   }
 }
 
-void debug(String msg) {
-  if (enableDebugLogging || (logLevel != null && logLevel.toInteger() == 1)) {
-    log.debug msg
-  }
+Boolean debugOk(){
+  String ll= (String)settings.logLevel
+  return  (ll != sNL && ll.toInteger() == i1) || (Boolean)settings.enableDebugLogging
 }
 
-void info(String msg) {
-  if (enableDebugLogging || logLevel == null
-      || (logLevel.toInteger() >= 1 && logLevel.toInteger() < 3)) {
-    log.info msg
-  }
+void debug(String msg) { if ( debugOk()) { logDebug msg } }
+
+Boolean infoOk() {
+  String ll= (String)settings.logLevel
+  return  (ll == sNL
+          || (ll.toInteger() >= i1 && ll.toInteger() < i3)
+          || (Boolean)settings.enableDebugLogging)
 }
 
-void warn(String msg) {
-  if (enableDebugLogging || logLevel == null || logLevel.toInteger() > 0) {
-     log.warn msg
+void info(String msg) { if ( infoOk()){ logInfo msg } }
+
+Boolean warnOk(){
+  String ll= (String)settings.logLevel
+  return (ll == sNL || ll.toInteger() > iZ
+          || (Boolean)settings.enableDebugLogging)
+}
+
+void warn(String msg) { if (warnOk()) { logWarn msg } }
+
+@Field static final String sBLK         = ''
+@Field static final String sSPACE         = ' '
+@Field static final String sCLRRED        = 'red'
+@Field static final String sCLRGRY        = 'gray'
+@Field static final String sCLRORG        = 'orange'
+@Field static final String sLINEBR        = '<br>'
+
+private void logDebug(String msg) { log.debug logPrefix(msg, "purple") }
+private void logInfo(String msg) { log.info sSPACE + logPrefix(msg, "#0299b1") }
+private void logTrace(String msg) { log.trace logPrefix(msg, sCLRGRY) }
+private void logWarn(String msg) { log.warn sSPACE + logPrefix(msg, sCLRORG) }
+
+void logError(String msg, ex=null) {
+  log.error logPrefix(msg, sCLRRED)
+  String a; a=sNL
+  try {
+    if (ex) a = getExceptionMessageWithLine(ex)
+  } catch (ignored) {}
+  if(a) log.error logPrefix(a, sCLRRED)
+}
+
+@CompileStatic
+static String logPrefix(String msg, String color = sNL) {
+  return span("Sleep Number App (v" + appVersionFLD + ") | ", sCLRGRY) + span(msg, color)
+}
+
+@CompileStatic
+static String span(String str, String clr=sNL, String sz=sNL, Boolean bld=false, Boolean br=false) { return str ? "<span ${(clr || sz || bld) ? "style='${clr ? "color: ${clr};" : sBLK}${sz ? "font-size: ${sz};" : sBLK}${bld ? "font-weight: bold;" : sBLK}'" : sBLK}>${str}</span>${br ? sLINEBR : sBLK}" : sBLK }
+
+
+
+
+
+// wrappers
+
+Long wnow() { return (Long)now() }
+Object wparseJson(String a) { return parseJson(a) }
+Boolean wtimeOfDayIsBetween(Date s, Date st, Date v) { return (Boolean)timeOfDayIsBetween(s,st,v) }
+Date wtoDateTime(String t) { return (Date)toDateTime(t) }
+void wrunIn(Long t, String meth, Map options=null){ runIn(t,meth,options) }
```

Comment:
what's the advantage of having a wrapper for this?

Reply:
I makes code type checking in IDE (I use IntelliJ) work as built-ins are otherwise unknown.  

It can make resulting code smaller as the JVM does not need to box and unbox the unknown types.